SparseSeries accepts scipy.sparse.spmatrix in constructor

kernc opened the pull request:
- [X] closes #15634
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [X] whatsnew entry

kernc committed:
PERF: avoid unnecessary array copy

kernc committed:
ENH: SparseArray constructor supports 1d scipy.sparse.spmatrix input

kernc committed:
ENH: Series constructor supports 1d scipy.sparse.spmatrix input

kernc committed:
ENH: SparseDataFrame supports scipy.sparse.spmatrix in setitem

kernc committed:
DOC: Document scipy sparse matrix accepted in SparseSeries constructor

jreback commented on doc/source/whatsnew/v0.20.2.txt:
```diff
@@ -25,6 +25,9 @@ Enhancements
   :ref:`here <groupby.ngroup>`.
 
+
+- ``SparseSeries`` and ``SparseArray`` now support 1d ``scipy.sparse.spmatrix`` in constructor. Additionally, ``SparseDataFrame`` can be assigned columns of ``scipy.sparse.spmatrix``; see :ref:`here <sparse.scipysparse_series>`. (:issue:`15634`)
```
will be for 0.21.0

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -433,6 +433,15 @@ def __getitem__(self, key):
+            else:
+                # 2d; make it iterable
+                value = list(value.tocsc().T)
+        super().__setitem__(key, value)
```
use the fully qualified call

jreback commented on pandas/core/sparse/series.py:
```diff
@@ -722,6 +726,9 @@ def combine_first(self, other):
     def to_coo(self, row_levels=(0, ), column_levels=(1, ), sort_labels=False):
         """
+        DEPRECATED; instead, make a SparseSeries with a two-level index,
+        unstack it, then use .to_coo() on the resulting SparseDataFrame.
```
use the deprecated sphinx directive (I think we are changing these all over)

jreback commented on pandas/core/sparse/series.py:
```diff
@@ -779,6 +786,9 @@ def to_coo(self, row_levels=(0, ), column_levels=(1, ), sort_labels=False):
     def from_coo(cls, A, dense_index=False):
         """
+        DEPRECATED; instead, pass 1d scipy.sparse matrices directly into
+        SparseSeries constructor, and 2d into SparseDataFrame constructor.
```
same

jreback reviewed (changes requested):
just a quick look. this is probably going to be tricky to make work.

kernc committed:
fixup! DOC: Document scipy sparse matrix accepted in SparseSeries constructor

kernc committed:
fixup! ENH: SparseDataFrame supports scipy.sparse.spmatrix in setitem

kernc commented:
> this is probably going to be tricky to make work.

Why do you think so? I does seem to work for the moment. 😃

kernc committed:
Fix tests.sparse.test_frame.TestSparseDataFrame.test_setitem_spmatrix

In Python 2, inner-block anonymous exception seems to overwrite the
outer-block anonymous exception. We're supposed to re-raise the latter.

kernc committed:
Fix CirleCI test on ancient SciPy

jreback commented on doc/source/sparse.rst:
```diff
@@ -213,9 +213,28 @@ To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you
+   sdf[['z', 'w']] = sp_arr[:, [7, 8]]
+   sdf.iloc[:, -5:]
+
+Below interface is deprecated.
```
say that this is deprecated in 0.21.0

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -25,6 +25,10 @@ New features
 - Added `__fspath__` method to :class`:pandas.HDFStore`, :class:`pandas.ExcelFile`,
   and :class:`pandas.ExcelWriter` to work properly with the file system path protocol (:issue:`13823`)
 
+- ``SparseSeries`` and ``SparseArray`` now support 1d ``scipy.sparse.spmatrix`` in constructor.
```
in the constructor

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -25,6 +25,10 @@ New features
   and :class:`pandas.ExcelWriter` to work properly with the file system path protocol (:issue:`13823`)
 
+- ``SparseSeries`` and ``SparseArray`` now support 1d ``scipy.sparse.spmatrix`` in constructor.
+  Additionally, ``SparseDataFrame`` can be assigned columns of ``scipy.sparse.spmatrix``;
```
make this 2nd sentence a separate bullet point (you can use same issue on both of them, or 2nd one should be the PR number maybe)

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -433,6 +433,16 @@ def __getitem__(self, key):
+                                    kind=self._default_kind)
+            else:
+                # 2d; make it iterable
+                value = list(value.tocsc().T)
```
does this materialize?

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -540,6 +540,37 @@ def test_setitem_array(self):
+        spm = csr_matrix(np.arange(len(sdf))).T
+        sdf['X'] = spm
+        assert _equal(sdf[['X']].to_coo(), spm)
+
```
this comparision on the scipy side is fine, but also let's compare with assert_sparse_series/frame_equal

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -540,6 +540,37 @@ def test_setitem_array(self):
+
+        # 1d row -- changing series contents not yet supported
+        spm = csr_matrix(np.arange(sdf.shape[1], dtype=float))
+        idx = np.zeros(sdf.shape[0], dtype=bool)
```
can you test with ``.loc/.iloc`` as well (might already be another issue about this, if not and its too complicated for here, then create a new issue)

jreback reviewed (changes requested):
looks pretty good. a couple of comments.

jreback commented:
can you rebase / update

jreback commented:
@kernc if you have time can you rebase / update

kernc commented:
For some reason (probably `assert_raises` in `test_setitem_spmatrix()`), I had a while ago decided to first implement `setitem` on `SparseSeries` that I have 80% done waiting in a feature branch somewhere. If you insist, I can push this first, however.

jreback commented:
oh we can do the other first 
was just looking thru open prs 

what was that number?

kernc commented:
Right, it's https://github.com/pandas-dev/pandas/pull/17785.

jreback commented:
closing as stale. if you want to continue working, pls ping.
